examples/platform/nrfconnect/util/DFUOverSMP.cpp

```diff
@@ -0,0 +1,131 @@
+
+    bt_conn_cb_register(&mBleConnCallbacks);
+
+    restartAdvertisingCallback = startAdvertisingCb;
+}
+
+void DFUOverSMP::ConfirmNewImage()
+{
+    // Check if the image is run in the REVERT mode and eventually
+    // confirm it to prevent reverting on the next boot.
+    if (mcuboot_swap_type() == BOOT_SWAP_TYPE_REVERT)
+    {
+        if (boot_write_img_confirmed())
+        {
+            ChipLogError(DeviceLayer, "Confirming firmware image failed, it will be reverted on the next boot.");
+        }
+        else
+        {
+            ChipLogProgress(DeviceLayer, "New firmware image confirmed.");
+        }
+    }
+}
+
+int DFUOverSMP::UploadConfirmHandler(uint32_t offset, uint32_t size, void * arg)
+{
+    // For now just print update progress and confirm data chunk without any additional checks.
+    ChipLogProgress(DeviceLayer, "Software update progress %d B / %d B", offset, size);
+
+    return 0;
+}
+
+void DFUOverSMP::StartServer()
+{
+    if (!mIsEnabled)
+    {
+        mIsEnabled = true;
+        smp_bt_register();
+
+        ChipLogProgress(DeviceLayer, "Enabled software update");
+
+        // Start SMP advertising only in case CHIPoBLE advertising is not working.
+        if (!chip::DeviceLayer::ConnectivityMgr().IsBLEAdvertisingEnabled())
+            StartBLEAdvertising();
+    }
+    else
+    {
+        ChipLogProgress(DeviceLayer, "Software update is already enabled");
+    }
+}
+
+void DFUOverSMP::StartBLEAdvertising()
+{
+    const char * deviceName = bt_get_name();
+    const uint8_t advFlags  = BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR;
+
+    bt_data ad[] = { BT_DATA(BT_DATA_FLAGS, &advFlags, sizeof(advFlags)),
+                     BT_DATA(BT_DATA_NAME_COMPLETE, deviceName, static_cast<uint8_t>(strlen(deviceName))) };
+
+    int rc;
+    bt_le_adv_param advParams = BT_LE_ADV_PARAM_INIT(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_ONE_TIME, BT_GAP_ADV_FAST_INT_MIN_2,
+                                                     BT_GAP_ADV_FAST_INT_MAX_2, nullptr);
+
+    rc = bt_le_adv_stop();
+    if (rc)
+    {
+        ChipLogError(DeviceLayer, "SMP advertising stop failed (rc %d)", rc);
+    }
+
+    rc = bt_le_adv_start(&advParams, ad, ARRAY_SIZE(ad), NULL, 0);
+    if (rc)
+    {
+        ChipLogError(DeviceLayer, "SMP advertising start failed (rc %d)", rc);
+    }
+    else
+    {
+        ChipLogProgress(DeviceLayer, "Started SMP service BLE advertising");
+    }
+}
+
+void DFUOverSMP::OnBleDisconnect(struct bt_conn * conId, uint8_t reason)
+{
+    // After BLE disconnect SMP advertising needs to be restarted. Before making it ensure that BLE disconnect was not triggered by
+    // closing CHIPoBLE service connection (in that case CHIPoBLE advertising needs to be restarted).
+    if (!chip::DeviceLayer::ConnectivityMgr().IsBLEAdvertisingEnabled() &&
+        chip::DeviceLayer::ConnectivityMgr().NumBLEConnections() == 0)
+    {
+        sDFUOverSMP.restartAdvertisingCallback();
```

Comment:
Shouldn't we check whether SMP is enabled? Preferable it should happen in `StartBLEAdvertising`. Also, ideally calls to `ConnectivityMgr()` in this context should be guarded by `LockChipStack()/UnlockChipStack` as this code doesn't run on the CHIP thread.